arangod/Graph/Enumerators/WeightedTwoSidedEnumerator.cpp

```diff
@@ -245,68 +249,77 @@ auto WeightedTwoSidedEnumerator<QueueType, PathStoreType, ProviderType,
+    ValidationResult res = _validator.validatePath(step);
+    TRI_ASSERT(!res.isFiltered() && !res.isPruned());
+
+    matchPathWeight = other.matchResultsInShell(step, candidates, _validator);
   }
 
   _provider.expand(step, previous, [&](Step n) -> void {
-    if (!hasBeenVisited(n)) {
-      verifyStep(std::move(n));
+    ValidationResult res = _validator.validatePath(n);
+
+    if (!res.isFiltered() && !res.isPruned()) {
+      // Add the step to our shell
+      _queue.append(std::move(n));
     }
   });
-}
 
-template<class QueueType, class PathStoreType, class ProviderType,
-         class PathValidator>
-void WeightedTwoSidedEnumerator<
-    QueueType, PathStoreType, ProviderType,
-    PathValidator>::Ball::testDepthZero(Ball& other,
-                                        CandidatesMap& candidates) {
-  // TODO: Check - is this even required?
-  /*for (auto const& step : _shell) {
-    other.matchResultsInShell(step, candidates, _validator);
-  }*/
+  return matchPathWeight;
 }
 
 template<class QueueType, class PathStoreType, class ProviderType,
          class PathValidator>
 auto WeightedTwoSidedEnumerator<QueueType, PathStoreType, ProviderType,
                                 PathValidator>::Ball::
-    matchResultsInShell(Step const& otherStep, CandidatesMap& candidates,
+    matchResultsInShell(Step const& otherStep, CandidatesStore& candidates,
                         PathValidator const& otherSideValidator) -> double {
-  auto position = _visitedNodes.at(otherStep.getVertex().getID());
-  auto ourStep = _interior.getStepReference(position);
-
-  // TODO: Check Prune + Filter - Currently, not supported, but could be (!) as
-  //  a new feature.
-  // auto res = _validator.validatePath(ourStep, otherSideValidator);
-  // TRI_ASSERT(!res.isFiltered());
-  // if (!res.isFiltered()) {
-  // }
-  double fullPathLength = ourStep.getWeight() + otherStep.getWeight();
-  if (_direction == FORWARD) {
-    candidates.emplace(fullPathLength, std::make_pair(ourStep, otherStep));
-  } else {
-    candidates.emplace(fullPathLength, std::make_pair(otherStep, ourStep));
+  double fullPathWeight = -1.0;
+  auto positions = _visitedNodes.at(otherStep.getVertex().getID());
+
+  for (auto const& position : positions) {
+    auto ourStep = _interior.getStepReference(position);
+
+    auto res = _validator.validatePath(ourStep, otherSideValidator);
+    if (res.isFiltered() || res.isPruned()) {
+      // This validator e.g. checks for path uniqueness violations.
+      continue;
+    }
+
+    double nextFullPathWeight = ourStep.getWeight() + otherStep.getWeight();
+
+    if (_direction == FORWARD) {
+      candidates.append(
+          std::make_tuple(nextFullPathWeight, ourStep, otherStep));
+    } else {
+      candidates.append(
+          std::make_tuple(nextFullPathWeight, otherStep, ourStep));
+    }
+
+    if (fullPathWeight == -1.0 || fullPathWeight > nextFullPathWeight) {
```

Comment:
```suggestion
    if (fullPathWeight < 0.0 || fullPathWeight > nextFullPathWeight) {
```
For floating point reasons, even if I think -1.0 is relatively safe.